plugin/trino-loki/src/test/java/io/trino/plugin/loki/TestLokiIntegration.java

```diff
@@ -0,0 +1,200 @@
+                          '{test="labels_complex"}',
+                          TIMESTAMP '%s',
+                          TIMESTAMP '%s'
+                        ))
+                        GROUP BY labels['service']
+                        """, timestampFormatter.format(start), timestampFormatter.format(end)),
+                "VALUES ('one', 2.0), ('two', 1.0)");
+    }
+
+    @Test
+    public void testSelectTimestamp()
+            throws Exception
+    {
+        DateTimeFormatter isoTimestampFormatter = DateTimeFormatter.ofPattern("uuuu-MM-dd'T'HH:mm:ss.SSS'Z'").withZone(ZoneOffset.UTC);
+        Instant start = Instant.now().truncatedTo(ChronoUnit.DAYS).minus(Duration.ofHours(12));
+        Instant end = start.plus(Duration.ofHours(4));
+        Instant firstLineTimestamp = start.plus(Duration.ofMinutes(20)).truncatedTo(ChronoUnit.MILLIS);
+
+        this.client.pushLogLine("line 1", firstLineTimestamp, ImmutableMap.of("test", "select_timestamp_query"));
+        this.client.pushLogLine("line 2", firstLineTimestamp.plus(Duration.ofMinutes(5)), ImmutableMap.of("test", "select_timestamp_query"));
+        this.client.pushLogLine("line 3", firstLineTimestamp.plus(Duration.ofMinutes(5)), ImmutableMap.of("test", "select_timestamp_query"));
+        this.client.flush();
+        assertQuery(String.format("""
+                        SELECT
+                          -- H2 does not support TIMESTAMP WITH TIME ZONE so cast to VARCHAR
+                          to_iso8601(timestamp), value
+                        FROM
+                        TABLE(system.query_range(
+                         '{test="select_timestamp_query"}',
+                         TIMESTAMP '%s',
+                         TIMESTAMP '%s'
+                        ))
+                        ORDER BY timestamp
+                        LIMIT 1
+                        """, timestampFormatter.format(start), timestampFormatter.format(end)),
+                String.format("VALUES ('%s', 'line 1')", isoTimestampFormatter.format(firstLineTimestamp)));
+
+        assertQuery(String.format("""
+                        SELECT
+                          -- H2 does not support TIMESTAMP WITH TIME ZONE so cast to VARCHAR
+                          to_iso8601(timestamp), value
+                        FROM
+                        TABLE(system.query_range(
+                         'count_over_time({test="select_timestamp_query"}[1h])',
+                         TIMESTAMP '%s',
+                         TIMESTAMP '%s'
+                        ))
+                        ORDER BY timestamp ASC
+                        LIMIT 1
+                        """, timestampFormatter.format(start), timestampFormatter.format(end)),
+                // Since the start is always noon the previous day we know how the steps align.
+                String.format("VALUES ('%s', 1.0)", isoTimestampFormatter.format(start.plus(Duration.ofMinutes(20)).plus(Duration.ofSeconds(45)))));
```

Comment:
What;s the reason for 45 seconds ?

Reply:
That's just how the steps are aligning. Without passing a [step](https://grafana.com/docs/loki/latest/reference/loki-http-api/#query-logs-within-a-range-of-time) it's hard to control.